src/controller.cpp

```diff
@@ -36,6 +36,14 @@ bool have_same_sign(T val1, T val2)
   return std::signbit(val1) == std::signbit(val2);
 }
 
+// Floating-point aware version of std::clamp(). fclamp() is to std::clamp() what std::fmin() and
+// std::fmax() are to std::min() and std::max().
+template <typename T>
+T fclamp(T value, T low, T high)
+{
+  return std::fmax(low, std::fmin(value, high));
+}
```

Comment:
Why didn't `std::clamp` not work for this case?

Reply:
That depends. In the original code `fmin()` and `fmax()` were used:
```cpp
t = fmax(0.0, fmin(1.0, t));
```

These functions explicitly deal with NAN values while `min()` and `max()` do not. I assumed this was a conscious choice by the original implementer.

If so, `std::clamp()` would not suffice.

If not, then I could change to `std::clamp()`.

Reply:
See https://en.cppreference.com/w/cpp/numeric/math/fmin

Reply:
Ah OK, you didn't want to change behavior of the code. But I think the original implementer didn't take paths that contain NaN values into account, so my preference is to use `std::clamp`. I'm curious what other reviewers think of this

Reply:
If paths contain NaN values, we should do an early check for that and reject the path, because I assume other paths of the code didn't take NaNs into account

Reply:
> If paths contain NaN values, we should do an early check for that and reject the path, because I assume other paths of the code didn't take NaNs into account

I agree with you, Ramon, but I guess that would be for another PR, to keep this one a clean refactor. I've raised issues for a few of these behavior/functionality improvements already, but I guess there is room for more.

Reply:
https://coliru.stacked-crooked.com/view?id=0b50b3c84a79727d

Seems like NAN input returns NAN output?

Reply:
Yes, but this is not the case for the `fmax` and `fmin`
> If one of the two arguments is NaN, the value of the other argument is returned
If the input is NaN, the current code returns `high` and this stays the same after the refactor

Reply:
> If one of the two arguments is NaN, the value of the other argument is returned

Ah yak. I'm quite sure this should not be the case. Let me look into this briefly and we can simply use `std::clang` here instead of opening issues for it.

Reply:
Verified, `std::clang` is better even.

Reply:
@Timple @Rayman fixed!